src/main/java/minijava/ir/optimize/Optimizer.java

```diff
@@ -52,23 +52,26 @@ static void optimize() {
             .dependsOn(controlFlowOptimizer, floatInTransformation)
             .build();
 
+    ProgramMetrics metrics = ProgramMetrics.analyse(firm.Program.getGraphs());
+    Inliner inliner = new Inliner(metrics, true);
     while (true) {
       for (Graph graph : firm.Program.getGraphs()) {
         perGraphFramework.optimizeUntilFixedpoint(graph);
       }
 
       // Here comes the interprocedural stuff... This is method is really turning into a mess
       boolean hasChanged = false;
-      ProgramMetrics metrics = ProgramMetrics.analyse(firm.Program.getGraphs());
-      Optimizer inliner = new Inliner(metrics);
       for (Graph graph : firm.Program.getGraphs()) {
         hasChanged |= inliner.optimize(graph);
         unreachableCodeRemover.optimize(graph);
-        metrics.updateGraphInfo(graph);
         Cli.dumpGraphIfNeeded(graph, "after-inlining");
       }
       if (!hasChanged) {
-        break;
+        if (inliner.onlyLeafs) {
+          inliner = new Inliner(metrics, false);
```

Comment:
There are certainly many ways to do this in an object oriented fashion.
Anyhow - as you said it won't make much of a difference, do you want me to change it or not?